tests/sub_group/sub_group_api.cpp

```diff
@@ -0,0 +1,368 @@
+                    [](bool index_is_seen) { return index_is_seen; }));
+
+  // check that all work-items in the same sub-group report the same sub-group
+  // size. use REQUIRE as sub_group_sizes is used by next checks
+  // ---------------------------------------------------------------------------
+
+  // for each sub-group, the number of work-items it contains
+  std::vector<size_t> sub_group_sizes(sub_group_count, special_val);
+  // populate and check sub_group_sizes by iterating over the work-items
+  for (size_t i = 0; i < linear_execution_range; i++) {
+    const size_t wg_id = results[get_idx(i, method::group_group_linear_id)];
+    const size_t local_sg_id = results[get_idx(i, method::group_id)];
+    // #sub-groups before this work-group + sub-group index in this work-group
+    //  = global sub-group index
+    const size_t global_sg_id = work_group_offsets[wg_id] + local_sg_id;
+    const size_t sub_group_size = results[get_idx(i, method::local_range)];
+    if (sub_group_sizes[global_sg_id] == special_val) {
+      sub_group_sizes[global_sg_id] = sub_group_size;
+    } else {
+      // if a previous work-item from the same sub-group already set the value,
+      // check that it is equal
+      REQUIRE((sub_group_size == sub_group_sizes[global_sg_id]));
+    }
+  }
+  // check that all elements were assigned a value, meaning that
+  // all unique sub-groups are processed
+  REQUIRE(std::all_of(sub_group_sizes.begin(), sub_group_sizes.end(),
+                      [=](size_t size) { return size != special_val; }));
+
+  // check that each sub-group has as many work-items are reported previously
+  // ---------------------------------------------------------------------------
+
+  // for each sub-group, the number of work-items before it
+  std::vector<size_t> sub_group_offsets(sub_group_count);
+  std::exclusive_scan(sub_group_sizes.begin(), sub_group_sizes.end(),
+                      sub_group_offsets.begin(), 0);
+
+  // the number of work-items across all sub-groups
+  size_t item_count =
+      std::reduce(sub_group_sizes.begin(), sub_group_sizes.end(), 0);
+  REQUIRE(item_count == linear_execution_range);
+
+  // for each item, whether there exists an item with that item id
+  std::vector<bool> item_seen(item_count, false);
+  // populate item_seen
+  for (size_t i = 0; i < linear_execution_range; i++) {
+    const size_t wg_id = results[get_idx(i, method::group_group_linear_id)];
+    const size_t local_sg_id = results[get_idx(i, method::group_id)];
+    // #sub-groups before this work-group + sub-group index in this work-group
+    //  = global sub-group index
+    const size_t global_sg_id = work_group_offsets[wg_id] + local_sg_id;
+    const size_t item_id = results[get_idx(i, method::local_id)];
+    // #items before this sub-group + local item index in this sub-group
+    //  = global item index
+    const size_t global_item_id = sub_group_offsets[global_sg_id] + item_id;
+    item_seen[global_item_id] = true;
```

Comment:
Same here. What happens if the id are completely wrong?

Reply:
Same solution.